src/vsr/replica.zig

```diff
@@ -5319,6 +5390,46 @@ pub fn ReplicaType(
             assert(reconfiguration_request.result != .reserved);
         }
 
+        fn pulse_inject(self: *Self, operation: StateMachine.Operation) void {
+            assert(self.status == .normal);
+            assert(self.primary());
+            assert(!self.pipeline.queue.full());
+            assert(!self.pipeline.queue.pulse_in_progress());
+
+            const message = self.message_bus.get_message(.prepare);
+            defer self.message_bus.unref(message);
+
+            const prepare_timestamp = self.state_machine.prepare_timestamp;
+            const latest_entry = self.journal.header_with_op(self.op).?;
+            message.header.* = Header.Prepare{
+                .cluster = self.cluster,
+                .size = @sizeOf(Header),
+                .view = self.view,
+                .release = self.release,
+                .command = .prepare,
+                .replica = self.replica,
+                .parent = latest_entry.checksum,
+                .request_checksum = 0,
+                .checkpoint_id = self.checkpoint_id_now(),
+                .op = self.op + 1,
+                .commit = self.commit_max,
+                .timestamp = prepare_timestamp,
+                .operation = vsr.Operation.from(StateMachine, operation),
+                .client = 0,
+                .request = 0,
+            };
+
+            assert(message.body().len == 0);
+            const sector_ceil = vsr.sector_ceil(message.header.size);
+            assert(message.header.size < sector_ceil);
+            @memset(message.buffer[message.header.size..sector_ceil], 0);
+
+            message.header.set_checksum_body(message.body());
+            message.header.set_checksum();
+
+            self.push_prepare(message);
+        }
```

Comment:
This duplicated logic makes me nervous; I'm wondering how we can share more with the standard prepare code path (`Replica.primary_pipeline_prepare()`).

What if instead of generating a `prepare`, we generated a `request` (which is simpler), then pass that request to `primary_pipeline_prepare`? That would also remove the `self.state_machine.prepare_timestamp` update from `on_pulse_timeout()`, and remove the need for the `checkpoint_id_now` helper.

Reply:
The main reason was to avoid recursion calling `primary_pipeline_prepare`, and to avoid skip calling `pulse_operation` for those requests. I'll refactor that.